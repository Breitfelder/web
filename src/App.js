import React, { Component } from 'react';
import styled from 'styled-components';
import 'normalize.css';
import SplitPane from 'react-split-pane';

import Header from './components/Header';
import Editor from './components/Editor';
import UASTViewer from './components/UASTViewer';
import languages from './languages';

const Wrap = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  display: flex;
  flex-direction: column;
`

const Content = styled.div`
  display: flex;
  flex-direction: row;
  position: relative;
  height: 100%;
`

const initialState = {
  languages: Object.assign({
    auto: { name: '(auto)' },
  }, languages),
  // babelfish tells us which language is active at the moment, but it
  // won't be used unless the selectedLanguage is auto.
  actualLanguage: 'python',
  loading: false,
  // this is the language that is selected by the user. It overrides the
  // actualLanguage except when it is 'auto'.
  selectedLanguage: 'auto',
  code: '',
  ast: undefined,
  userHasTyped: false,
};

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = Object.assign({}, initialState);
  }

  onLanguageChanged(e) {
    let selectedLanguage = e.target.value;
    if (!this.hasLanguage(selectedLanguage)) {
      selectedLanguage = 'auto';
    }
    this.setState({ selectedLanguage });
  }

  hasLanguage(lang) {
    return this.state.languages.hasOwnProperty(lang);
  }

  onRunParser(e) {
    throw new Error('run parser not implemented yet!');
  }

  onNodeSelected(from, to) {
    return this.refs.editor.selectCode(from, to);
  }

  onCursorChanged(start, end) {
    this.refs.viewer.selectNodes(start, end);
  }

  onCodeChange(code) {
    this.setState({ code, userHasTyped: true });
  }

  render() {
    const { innerWidth: width } = window;
    const { 
      languages, 
      selectedLanguage, 
      code, 
      ast, 
      loading, 
      actualLanguage,
      userHasTyped,
    } = this.state;

    return (
      <Wrap>
        <Header
          languages={languages}
          selectedLanguage={selectedLanguage}
          actualLanguage={actualLanguage}
          onLanguageChanged={e => this.onLanguageChanged(e)}
          onRunParser={e => this.onRunParser(e)}
          userHasTyped={userHasTyped}
          loading={loading} />

        <Content>
          <SplitPane 
            split='vertical' 
            minSize={width * 0.25} 
            defaultSize='50%' 
            maxSize={width * 0.75}>

            <Editor
              ref='editor'
              code={code}
              languageMode={languages[actualLanguage].mode}
              onChange={code => this.onCodeChange(code)}
              onCursorChanged={editor => this.onCursorChanged(editor)} />

            <UASTViewer 
              ref='viewer'
              onNodeSelected={(from, to) => this.onNodeSelected(from, to)}
              ast={ast}
              loading={loading} />
          </SplitPane>
        </Content>
      </Wrap>
    );
  }
}
