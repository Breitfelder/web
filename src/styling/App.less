.bblfsh-app {
    &__wrap {
        position: absolute;
        top: 0;
        left: 0;
        bottom: 0;
        right: 0;
        display: flex;
        flex-direction: column;
    }

    &__content {
        display: flex;
        height: 100%;
        flex-direction: row;
        position: relative;
    }

    &__right-panel {
        display: flex;
        height: 100%;
        flex-direction: column;
        position: relative;
        background: #fff;
    }
}

.ReactCodeMirror,
.CodeMirror {
    height: 100%;
}

.CodeMirror {
    font-family: @font-family-code;
    font-size: @font-size-large;
}
