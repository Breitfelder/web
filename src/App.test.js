import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import 'jest-styled-components';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

async function renderApp(state) {
  fetch.mockResponse(
    JSON.stringify([
      { id: 'python', name: 'Python', url: 'python-driver' },
      { id: 'java', name: 'Java', url: 'java-driver' }
    ])
  );

  const elem = document.createElement('div');
  const component = ReactDOM.render(<App />, elem);
  if (state) {
    component.setState(state);
  }

  if (component.loaded) {
    await component.loaded;
  }

  return component;
}

it('renders without crashing', () => {
  const wrapper = shallow(<App />);
  expect(toJson(wrapper)).toMatchStyledComponentsSnapshot();
});

it('removes the error when onErrorRemoved is called', async () => {
  const wrapper = await renderApp({ errors: ['1', '2', '3'] });
  wrapper.onErrorRemoved(1);
  expect(wrapper.state.errors).toEqual(['1', '3']);
});

it('changes the editor mode when it updates', async () => {
  const wrapper = await renderApp();
  expect(wrapper.refs.editor.document.getMode().name).toBe('python');

  wrapper.setState({ selectedLanguage: 'java' });
  expect(wrapper.refs.editor.document.getMode().name).toBe('clike');
});

it('onLanguageChanged changes the selected language', async () => {
  const wrapper = await renderApp();
  expect(wrapper.state.selectedLanguage).toBe('auto');

  wrapper.onLanguageChanged('java');
  expect(wrapper.state.selectedLanguage).toBe('java');

  wrapper.onLanguageChanged('does not exist');
  expect(wrapper.state.selectedLanguage).toBe('auto');
});

it('onNodeSelected clears the previously set mark and sets a new one', async () => {
  const clear = jest.fn();
  const wrapper = await renderApp();
  wrapper.mark = { initial: true, clear };
  wrapper.onNodeSelected({ line: 0, ch: 0 }, { line: 0, ch: 2 });

  expect(clear.mock.calls.length).toBe(1);
  expect(wrapper.mark.initial).toBeUndefined();
});

it('clearNodeSelection clears the current mark', async () => {
  const clear = jest.fn();
  const wrapper = await renderApp();
  wrapper.mark = { clear };
  wrapper.clearNodeSelection();

  expect(clear.mock.calls.length).toBe(1);
  expect(wrapper.mark).toBeNull();
});
