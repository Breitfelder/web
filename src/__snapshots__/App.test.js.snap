// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<styled.div>
  <Header
    actualLanguage="python"
    dirty={false}
    languages={
      Object {
        "auto": Object {
          "name": "(auto)",
        },
        "java": Object {
          "mode": "text/x-java",
          "name": "Java",
          "url": "#",
        },
        "python": Object {
          "mode": "python",
          "name": "Python",
          "url": "#",
        },
      }
    }
    loading={false}
    onLanguageChanged={[Function]}
    onRunParser={[Function]}
    selectedLanguage="auto"
  />
  <styled.div>
    <SplitPane
      allowResize={true}
      defaultSize="50%"
      maxSize={768}
      minSize={256}
      prefixer={
        Prefixer {
          "_browserInfo": Object {
            "browserName": "safari",
            "browserVersion": 7,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "name": "Safari",
            "osVersion": 7,
            "osversion": "7",
            "safari": true,
            "webkit": true,
            "windows": true,
            "x": true,
          },
          "_hasPropsRequiringPrefix": true,
          "_keepUnprefixed": false,
          "_metaData": Object {
            "browserName": "safari",
            "browserVersion": 7,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "keepUnprefixed": false,
            "requiresPrefix": Object {
              "alignContent": true,
              "alignItems": true,
              "alignSelf": true,
              "animation": true,
              "animationDelay": true,
              "animationDirection": true,
              "animationDuration": true,
              "animationFillMode": true,
              "animationIterationCount": true,
              "animationName": true,
              "animationPlayState": true,
              "animationTimingFunction": true,
              "appearance": true,
              "backdropFilter": true,
              "backfaceVisibility": true,
              "boxDecorationBreak": true,
              "breakAfter": true,
              "breakBefore": true,
              "breakInside": true,
              "clipPath": true,
              "columnCount": true,
              "columnFill": true,
              "columnGap": true,
              "columnRule": true,
              "columnRuleColor": true,
              "columnRuleStyle": true,
              "columnRuleWidth": true,
              "columnSpan": true,
              "columnWidth": true,
              "columns": true,
              "filter": true,
              "flex": true,
              "flexBasis": true,
              "flexDirection": true,
              "flexFlow": true,
              "flexGrow": true,
              "flexShrink": true,
              "flexWrap": true,
              "flowFrom": true,
              "flowInto": true,
              "fontKerning": true,
              "hyphens": true,
              "justifyContent": true,
              "mask": true,
              "maskBorder": true,
              "maskBorderMode": true,
              "maskBorderOutset": true,
              "maskBorderRepeat": true,
              "maskBorderSlice": true,
              "maskBorderSource": true,
              "maskBorderWidth": true,
              "maskClip": true,
              "maskComposite": true,
              "maskImage": true,
              "maskMode": true,
              "maskOrigin": true,
              "maskPosition": true,
              "maskRepeat": true,
              "maskSize": true,
              "maskType": true,
              "order": true,
              "perspective": true,
              "perspectiveOrigin": true,
              "regionFragment": true,
              "scrollSnapCoordinate": true,
              "scrollSnapDestination": true,
              "scrollSnapPointsX": true,
              "scrollSnapPointsY": true,
              "scrollSnapType": true,
              "shapeImageMargin": true,
              "shapeImageOutside": true,
              "shapeImageThreshold": true,
              "textDecorationColor": true,
              "textDecorationLine": true,
              "textDecorationSkip": true,
              "textDecorationStyle": true,
              "transform": true,
              "transformOrigin": true,
              "transformOriginX": true,
              "transformOriginY": true,
              "transformOriginZ": true,
              "transformStyle": true,
              "userSelect": true,
            },
          },
          "_requiresPrefix": Object {
            "alignContent": true,
            "alignItems": true,
            "alignSelf": true,
            "animation": true,
            "animationDelay": true,
            "animationDirection": true,
            "animationDuration": true,
            "animationFillMode": true,
            "animationIterationCount": true,
            "animationName": true,
            "animationPlayState": true,
            "animationTimingFunction": true,
            "appearance": true,
            "backdropFilter": true,
            "backfaceVisibility": true,
            "boxDecorationBreak": true,
            "breakAfter": true,
            "breakBefore": true,
            "breakInside": true,
            "clipPath": true,
            "columnCount": true,
            "columnFill": true,
            "columnGap": true,
            "columnRule": true,
            "columnRuleColor": true,
            "columnRuleStyle": true,
            "columnRuleWidth": true,
            "columnSpan": true,
            "columnWidth": true,
            "columns": true,
            "filter": true,
            "flex": true,
            "flexBasis": true,
            "flexDirection": true,
            "flexFlow": true,
            "flexGrow": true,
            "flexShrink": true,
            "flexWrap": true,
            "flowFrom": true,
            "flowInto": true,
            "fontKerning": true,
            "hyphens": true,
            "justifyContent": true,
            "mask": true,
            "maskBorder": true,
            "maskBorderMode": true,
            "maskBorderOutset": true,
            "maskBorderRepeat": true,
            "maskBorderSlice": true,
            "maskBorderSource": true,
            "maskBorderWidth": true,
            "maskClip": true,
            "maskComposite": true,
            "maskImage": true,
            "maskMode": true,
            "maskOrigin": true,
            "maskPosition": true,
            "maskRepeat": true,
            "maskSize": true,
            "maskType": true,
            "order": true,
            "perspective": true,
            "perspectiveOrigin": true,
            "regionFragment": true,
            "scrollSnapCoordinate": true,
            "scrollSnapDestination": true,
            "scrollSnapPointsX": true,
            "scrollSnapPointsY": true,
            "scrollSnapType": true,
            "shapeImageMargin": true,
            "shapeImageOutside": true,
            "shapeImageThreshold": true,
            "textDecorationColor": true,
            "textDecorationLine": true,
            "textDecorationSkip": true,
            "textDecorationStyle": true,
            "transform": true,
            "transformOrigin": true,
            "transformOriginX": true,
            "transformOriginY": true,
            "transformOriginZ": true,
            "transformStyle": true,
            "userSelect": true,
          },
          "_userAgent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.2 (KHTML, like Gecko) Safari/537.2",
          "prefixedKeyframes": "-webkit-keyframes",
        }
      }
      primary="first"
      split="vertical"
    >
      <Editor
        code=""
        languageMode="python"
        onChange={[Function]}
        onCursorChanged={[Function]}
      />
      <UASTViewer
        clearNodeSelection={[Function]}
        loading={false}
        onNodeSelected={[Function]}
      />
    </SplitPane>
  </styled.div>
</styled.div>
`;
